File: mem_probe/proto_entity_locator.py
"""Proto-based entity locator — uses dump.cs klass RVAs to find Zproto.Entity
instances directly, then walks AttrCollection → Attr → ByteString → varint
to extract HP / MaxHP / break-bar / etc.

Layout (from dump.cs):
    Entity           +0x10 long Uuid    +0x18 EEntityType EntType
                     +0x20 AttrCollection*
    AttrCollection   +0x10 long Uuid    +0x18 RepeatedField<Attr>* Attrs
    RepeatedField<T> +0x10 T[] array    +0x18 int count
    Attr             +0x10 int Id       +0x18 ByteString*
    ByteString       +0x10 bool isBytesOwner   +0x18 ReadOnlyMemory<byte>
    ReadOnlyMemory   +0x18 byte[] obj   +0x20 int start   +0x24 int length
    byte[]           +0x20 raw data start

Entity klass RVA discovered values (script.json from regenerated dump):
    Zproto.Entity         RVA = 0x96C44D8  → klass_ptr = GA_base + 0x96C44D8
    Zproto.AttrCollection RVA = 0x96A3F08
    Zproto.Attr           RVA = 0x96A36D8
"""
from __future__ import annotations

import os
import struct
import sys
import threading
import time
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

_HERE = os.path.dirname(os.path.abspath(__file__))
_SAO_AUTO = os.path.dirname(_HERE)
if _SAO_AUTO not in sys.path:
    sys.path.insert(0, _SAO_AUTO)

from mem_probe import cy_memscan as _cy

# Klass RVAs (relative to GameAssembly.dll base)
ENTITY_KLASS_RVA          = 0x96C44D8
ATTRCOLLECTION_KLASS_RVA  = 0x96A3F08
ATTR_KLASS_RVA            = 0x96A36D8
USERFIGHTATTR_KLASS_RVA   = 0x96A7C38
CHARSERIALIZE_KLASS_RVA   = 0x95F5E18

# Field offsets (from dump.cs)
ENTITY_UUID_OFF           = 0x10  # long
ENTITY_ENTTYPE_OFF        = 0x18  # EEntityType (i32)
ENTITY_ATTRS_OFF          = 0x20  # AttrCollection*

ATTRCOLL_UUID_OFF         = 0x10  # long
ATTRCOLL_ATTRS_OFF        = 0x18  # RepeatedField<Attr>*

# RepeatedField<T> stores `T[] array` and `int count`.
# Standard IL2CPP generic ref-type instance layout:
#   +0x00  klass_ptr
#   +0x08  monitor
#   +0x10  T[] array (ptr)
#   +0x18  int count
REPFIELD_ARRAY_OFF        = 0x10
REPFIELD_COUNT_OFF        = 0x18

# IL2CPP T[] array layout:
#   +0x00  klass_ptr
#   +0x08  monitor
#   +0x10  bounds
#   +0x18  max_length (i64)
#   +0x20  data start
ARRAY_DATA_OFF            = 0x20

ATTR_ID_OFF               = 0x10  # i32
ATTR_RAWDATA_OFF          = 0x18  # ByteString*

# ByteString stores the bytes via ReadOnlyMemory<byte>:
#   +0x10  bool isBytesOwner
#   +0x18  ReadOnlyMemory.object  (byte[] ref)
#   +0x20  ReadOnlyMemory.start   (i32)
#   +0x24  ReadOnlyMemory.length  (i32)
BYTESTR_OBJ_OFF           = 0x18
BYTESTR_START_OFF         = 0x20
BYTESTR_LENGTH_OFF        = 0x24

# AttrType IDs we care about (from packet_parser.AttrType)
ATTR_HP                   = 0x2C2E   # 11310 — current HP
ATTR_MAX_HP               = 0x2C38   # 11320 — max HP (rarely sent for monsters)
ATTR_NAME                 = 0x01
ATTR_TEMPLATE_ID          = 0x02     # placeholder, check packet_parser

# Monster-specific extended (from packet_parser.AttrType)
ATTR_MAX_EXTINCTION       = 440
ATTR_EXTINCTION           = 441
ATTR_BREAKING_STAGE       = 455
ATTR_MONSTER_SEASON_LEVEL = 462

ENT_TYPE_CHAR    = 0
ENT_TYPE_MONSTER = 1


def _decode_varint_i32(raw: bytes) -> int:
    """Decode a protobuf varint payload to signed int32. Pure-Python fallback
    when _sao_cy_packet isn't available; the real Cython kernel is faster but
    this discovery is one-shot so speed not critical here."""
    val = 0
    shift = 0
    for i, b in enumerate(raw):
        if i >= 10:
            break
        val |= (b & 0x7F) << shift
        if (b & 0x80) == 0:
            break
        shift += 7
    val32 = val & 0xFFFFFFFF
    if val32 >= 0x80000000:
        return val32 - 0x100000000
    return val32


@dataclass
class EntitySnapshot:
    addr: int            # Entity instance address
    uuid: int
    ent_type: int        # EEntityType
    attrs_coll_ptr: int  # AttrCollection*
    attrs: Dict[int, int] = field(default_factory=dict)  # attr_id -> decoded i32 value
    raw_attrs: List[Tuple[int, int]] = field(default_factory=list)  # [(attr_addr, attr_id), ...]


# ───────── GA module helpers ─────────

def _ga_base(pm) -> int:
    for m in pm.list_modules():
        if m.name.lower() == "gameassembly.dll":
            return m.base
    raise RuntimeError("GameAssembly.dll not found in modules")


def klass_ptr_for(pm, rva: int) -> int:
    return _ga_base(pm) + rva


# ───────── Klass instance scan ─────────

MAX_REGION_SIZE = 256 * 1024 * 1024


def find_klass_instances(pm, klass_ptr: int, *, n_workers: int = 8) -> List[int]:
    """Cython AVX2 full-heap scan for u64 == klass_ptr. Returns list of
    instance object addresses (each starts with the klass_ptr).
    """
    needle = int(klass_ptr) & 0xFFFFFFFFFFFFFFFF
    regions = [r for r in pm.iter_regions(only_readable=True, only_private=True)
               if r.size <= MAX_REGION_SIZE]
    out: List[int] = []
    lock = threading.Lock()

    def scan_one(region):
        buf = pm.read_bytes(region.base, region.size)
        if buf is None:
            return
        offs = _cy.find_aligned_u64(buf, needle)
        if not offs:
            return
        with lock:
            for off in offs:
                out.append(region.base + off)

    if n_workers > 1:
        with ThreadPoolExecutor(max_workers=n_workers) as ex:
            list(ex.map(scan_one, regions))
    else:
        for r in regions:
            scan_one(r)
    return out


# ───────── Entity field walking ─────────

def _read_repeated_field_attrs(pm, attrs_coll_ptr: int) -> List[int]:
    """Given AttrCollection*, return list of Attr* in its Attrs RepeatedField.
    Returns empty list on any read failure.
    """
    repfield_blob = pm.read_bytes(attrs_coll_ptr, 0x40)
    if not repfield_blob or len(repfield_blob) < ATTRCOLL_ATTRS_OFF + 8:
        return []
    repfield_ptr = int.from_bytes(
        repfield_blob[ATTRCOLL_ATTRS_OFF:ATTRCOLL_ATTRS_OFF + 8], "little")
    if not (0x10000 <= repfield_ptr <= 0x7FFFFFFFFFFF):
        return []
    rf_blob = pm.read_bytes(repfield_ptr, 0x40)
    if not rf_blob or len(rf_blob) < REPFIELD_COUNT_OFF + 4:
        return []
    array_ptr = int.from_bytes(
        rf_blob[REPFIELD_ARRAY_OFF:REPFIELD_ARRAY_OFF + 8], "little")
    count = int.from_bytes(
        rf_blob[REPFIELD_COUNT_OFF:REPFIELD_COUNT_OFF + 4], "little", signed=True)
    if count <= 0 or count > 1024 or not (0x10000 <= array_ptr <= 0x7FFFFFFFFFFF):
        return []
    # Read array elements in one RPM
    arr_total_size = ARRAY_DATA_OFF + count * 8
    arr_blob = pm.read_bytes(array_ptr, min(arr_total_size, 0x4000))
    if not arr_blob or len(arr_blob) < ARRAY_DATA_OFF:
        return []
    out: List[int] = []
    for i in range(count):
        off = ARRAY_DATA_OFF + i * 8
        if off + 8 > len(arr_blob):
            break
        ptr = int.from_bytes(arr_blob[off:off + 8], "little")
        if 0x10000 <= ptr <= 0x7FFFFFFFFFFF:
            out.append(ptr)
    return out


def _read_attr_id_and_rawbytes(pm, attr_ptr: int) -> Optional[Tuple[int, bytes]]:
    """Read Attr.{Id, RawData bytes}. Returns (id, raw_bytes) or None."""
    blob = pm.read_bytes(attr_ptr, ATTR_RAWDATA_OFF + 8)
    if not blob or len(blob) < ATTR_RAWDATA_OFF + 8:
        return None
    attr_id = int.from_bytes(blob[ATTR_ID_OFF:ATTR_ID_OFF + 4], "little", signed=True)
    bs_ptr = int.from_bytes(blob[ATTR_RAWDATA_OFF:ATTR_RAWDATA_OFF + 8], "little")
    if not (0x10000 <= bs_ptr <= 0x7FFFFFFFFFFF):
        return (attr_id, b"")
    bs_blob = pm.read_bytes(bs_ptr, BYTESTR_LENGTH_OFF + 4)
    if not bs_blob or len(bs_blob) < BYTESTR_LENGTH_OFF + 4:
        return (attr_id, b"")
    arr_ptr = int.from_bytes(bs_blob[BYTESTR_OBJ_OFF:BYTESTR_OBJ_OFF + 8], "little")
    start = int.from_bytes(bs_blob[BYTESTR_START_OFF:BYTESTR_START_OFF + 4], "little", signed=True)
    length = int.from_bytes(bs_blob[BYTESTR_LENGTH_OFF:BYTESTR_LENGTH_OFF + 4], "little", signed=True)
    if length < 0 or length > 256 or not (0x10000 <= arr_ptr <= 0x7FFFFFFFFFFF):
        return (attr_id, b"")
    # byte[] data starts at +0x20 of the array object
    raw = pm.read_bytes(arr_ptr + ARRAY_DATA_OFF + max(0, start), length)
    return (attr_id, raw or b"")


def read_entity_full(pm, entity_addr: int, *,
                     decode_attrs: bool = True,
                     attr_filter: Optional[set] = None
                     ) -> Optional[EntitySnapshot]:
    """Read a full Entity at entity_addr. Returns None if invalid."""
    body = pm.read_bytes(entity_addr, 0x40)
    if not body or len(body) < ENTITY_ATTRS_OFF + 8:
        return None
    uuid = int.from_bytes(body[ENTITY_UUID_OFF:ENTITY_UUID_OFF + 8], "little", signed=True)
    ent_type = int.from_bytes(body[ENTITY_ENTTYPE_OFF:ENTITY_ENTTYPE_OFF + 4], "little", signed=True)
    attrs_coll_ptr = int.from_bytes(body[ENTITY_ATTRS_OFF:ENTITY_ATTRS_OFF + 8], "little")
    snap = EntitySnapshot(addr=entity_addr, uuid=uuid, ent_type=ent_type,
                          attrs_coll_ptr=attrs_coll_ptr)
    if not decode_attrs or attrs_coll_ptr == 0:
        return snap
    if not (0x10000 <= attrs_coll_ptr <= 0x7FFFFFFFFFFF):
        return snap
    attr_ptrs = _read_repeated_field_attrs(pm, attrs_coll_ptr)
    for attr_ptr in attr_ptrs:
        result = _read_attr_id_and_rawbytes(pm, attr_ptr)
        if not result:
            continue
        attr_id, raw = result
        snap.raw_attrs.append((attr_ptr, attr_id))
        if attr_filter is not None and attr_id not in attr_filter:
            continue
        if not raw:
            continue
        # All known attrs we care about decode as varint i32
        snap.attrs[attr_id] = _decode_varint_i32(raw)
    return snap


def discover_all_entities(pm, *, ent_type_filter: Optional[int] = None,
                          decode_attrs: bool = True,
                          attr_filter: Optional[set] = None,
                          verbose: bool = True) -> List[EntitySnapshot]:
    """Find all live Zproto.Entity instances and decode their attributes."""
    t0 = time.time()
    klass_ptr = klass_ptr_for(pm, ENTITY_KLASS_RVA)
    if verbose:
        print(f"[proto-entity] Entity klass_ptr = 0x{klass_ptr:X}")
        print(f"[proto-entity] scanning Entity instances ...")
    inst_addrs = find_klass_instances(pm, klass_ptr)
    if verbose:
        print(f"[proto-entity] found {len(inst_addrs)} Entity instances "
              f"({time.time()-t0:.2f}s)")
    snapshots: List[EntitySnapshot] = []
    for addr in inst_addrs:
        snap = read_entity_full(pm, addr,
                                decode_attrs=decode_attrs,
                                attr_filter=attr_filter)
        if snap is None:
            continue
        if ent_type_filter is not None and snap.ent_type != ent_type_filter:
            continue
        # uuid==0 → pool-resident / cleared instance, skip
        if snap.uuid == 0:
            continue
        snapshots.append(snap)
    if verbose:
        print(f"[proto-entity] decoded {len(snapshots)} live entities "
              f"({time.time()-t0:.2f}s total)")
    return snapshots


__all__ = [
    "EntitySnapshot",
    "find_klass_instances", "klass_ptr_for",
    "read_entity_full", "discover_all_entities",
    "ENTITY_KLASS_RVA", "ATTRCOLLECTION_KLASS_RVA", "ATTR_KLASS_RVA",
    "ATTR_HP", "ATTR_MAX_HP", "ATTR_EXTINCTION", "ATTR_MAX_EXTINCTION",
    "ATTR_BREAKING_STAGE", "ENT_TYPE_CHAR", "ENT_TYPE_MONSTER",
]

File: C#/src/SaoAuto.MemProbe/Discovery/ProtoEntityLocator.cs
using System.Buffers.Binary;

namespace SaoAuto.MemProbe.Discovery;

/// <summary>
/// IL2CPP layout constants for Zproto.Entity / AttrCollection / Attr /
/// ByteString — mirrors the table at the top of
/// <c>mem_probe/proto_entity_locator.py</c>. Treat as a frozen contract:
/// regenerated dump.cs RVAs may shift between game builds, so callers
/// resolve klass pointers via <see cref="ProtoEntityLocator.KlassPtrFor"/>.
/// </summary>
public static class ProtoEntityLayout
{
    public const ulong EntityKlassRva         = 0x96C44D8;
    public const ulong AttrCollectionKlassRva = 0x96A3F08;
    public const ulong AttrKlassRva           = 0x96A36D8;

    public const int EntityUuidOff    = 0x10;
    public const int EntityEntTypeOff = 0x18;
    public const int EntityAttrsOff   = 0x20;

    public const int AttrCollUuidOff  = 0x10;
    public const int AttrCollAttrsOff = 0x18;

    public const int RepFieldArrayOff = 0x10;
    public const int RepFieldCountOff = 0x18;

    public const int ArrayDataOff     = 0x20;

    public const int AttrIdOff        = 0x10;
    public const int AttrRawDataOff   = 0x18;

    public const int ByteStrObjOff    = 0x18;
    public const int ByteStrStartOff  = 0x20;
    public const int ByteStrLengthOff = 0x24;

    public const int AttrHp              = 0x2C2E;
    public const int AttrMaxHp           = 0x2C38;
    public const int AttrMaxExtinction   = 440;
    public const int AttrExtinction      = 441;
    public const int AttrBreakingStage   = 455;
    public const int AttrMonsterSeasonLevel = 462;

    public const int EntTypeChar    = 0;
    public const int EntTypeMonster = 1;

    public const int MaxAttrCount   = 1024;
    public const int MaxRawAttrLen  = 256;

    // Conservative user-mode pointer range used by the Python locator.
    public const ulong PtrLowBound  = 0x10000UL;
    public const ulong PtrHighBound = 0x7FFFFFFFFFFFUL;

    public static bool IsPlausiblePtr(ulong p)
        => p >= PtrLowBound && p <= PtrHighBound;
}

/// <summary>One decoded Entity instance — port of Python <c>EntitySnapshot</c>.</summary>
public sealed record EntitySnapshot(
    ulong Address,
    long Uuid,
    int EntType,
    ulong AttrsCollPtr,
    IReadOnlyDictionary<int, int> Attrs,
    IReadOnlyList<(ulong AttrAddress, int AttrId)> RawAttrs);

/// <summary>
/// Walks Zproto.Entity → AttrCollection → RepeatedField&lt;Attr&gt; → Attr →
/// ByteString → varint payload, all via <see cref="IMemorySource"/>. This is
/// the deterministic CPU half of <c>mem_probe/proto_entity_locator.py</c>;
/// the heap-scan side (<c>find_klass_instances</c>) lives on top of
/// <c>CyMemScan.FindAlignedU64</c> and is intentionally not duplicated here.
/// </summary>
public static class ProtoEntityLocator
{
    /// <summary>Resolve a klass pointer = GameAssembly base + RVA.</summary>
    public static ulong KlassPtrFor(ulong gameAssemblyBase, ulong rva)
        => gameAssemblyBase + rva;

    /// <summary>
    /// Decode a protobuf varint payload to signed int32. Caps at 10 bytes
    /// (the standard varint envelope), matches the Python reference exactly
    /// including the 32-bit sign-extend wrap.
    /// </summary>
    public static int DecodeVarintI32(ReadOnlySpan<byte> raw)
    {
        ulong val = 0;
        int shift = 0;
        for (int i = 0; i < raw.Length && i < 10; i++)
        {
            byte b = raw[i];
            val |= (ulong)(b & 0x7F) << shift;
            if ((b & 0x80) == 0) break;
            shift += 7;
        }
        uint v32 = (uint)(val & 0xFFFFFFFFUL);
        return unchecked((int)v32);
    }

    /// <summary>
    /// Read AttrCollection*.Attrs (a RepeatedField&lt;Attr&gt;) and return the
    /// list of Attr* pointers. Returns an empty list on any read failure
    /// (touching unmapped or torn pages is normal during scanning).
    /// </summary>
    public static IReadOnlyList<ulong> ReadRepeatedFieldAttrs(IMemorySource pm, ulong attrsCollPtr)
    {
        var collBlob = pm.ReadBytes(attrsCollPtr, 0x40);
        if (collBlob is null || collBlob.Length < ProtoEntityLayout.AttrCollAttrsOff + 8)
            return Array.Empty<ulong>();

        ulong repFieldPtr = BinaryPrimitives.ReadUInt64LittleEndian(
            collBlob.AsSpan(ProtoEntityLayout.AttrCollAttrsOff, 8));
        if (!ProtoEntityLayout.IsPlausiblePtr(repFieldPtr))
            return Array.Empty<ulong>();

        var rfBlob = pm.ReadBytes(repFieldPtr, 0x40);
        if (rfBlob is null || rfBlob.Length < ProtoEntityLayout.RepFieldCountOff + 4)
            return Array.Empty<ulong>();

        ulong arrayPtr = BinaryPrimitives.ReadUInt64LittleEndian(
            rfBlob.AsSpan(ProtoEntityLayout.RepFieldArrayOff, 8));
        int count = BinaryPrimitives.ReadInt32LittleEndian(
            rfBlob.AsSpan(ProtoEntityLayout.RepFieldCountOff, 4));
        if (count <= 0 || count > ProtoEntityLayout.MaxAttrCount
            || !ProtoEntityLayout.IsPlausiblePtr(arrayPtr))
            return Array.Empty<ulong>();

        int wantBytes = ProtoEntityLayout.ArrayDataOff + count * 8;
        int readSize = Math.Min(wantBytes, 0x4000);
        var arrBlob = pm.ReadBytes(arrayPtr, readSize);
        if (arrBlob is null || arrBlob.Length < ProtoEntityLayout.ArrayDataOff)
            return Array.Empty<ulong>();

        var result = new List<ulong>(count);
        for (int i = 0; i < count; i++)
        {
            int off = ProtoEntityLayout.ArrayDataOff + i * 8;
            if (off + 8 > arrBlob.Length) break;
            ulong p = BinaryPrimitives.ReadUInt64LittleEndian(arrBlob.AsSpan(off, 8));
            if (ProtoEntityLayout.IsPlausiblePtr(p))
                result.Add(p);
        }
        return result;
    }

    /// <summary>
    /// Read Attr.{Id, RawData bytes}. Returns null on outright read failure;
    /// returns (id, empty) when the id is readable but the ByteString chain
    /// is torn / pool-resident (matches the Python "(attr_id, b'')" path).
    /// </summary>
    public static (int Id, byte[] Raw)? ReadAttrIdAndRawBytes(IMemorySource pm, ulong attrPtr)
    {
        var blob = pm.ReadBytes(attrPtr, ProtoEntityLayout.AttrRawDataOff + 8);
        if (blob is null || blob.Length < ProtoEntityLayout.AttrRawDataOff + 8)
            return null;

        int attrId = BinaryPrimitives.ReadInt32LittleEndian(
            blob.AsSpan(ProtoEntityLayout.AttrIdOff, 4));
        ulong bsPtr = BinaryPrimitives.ReadUInt64LittleEndian(
            blob.AsSpan(ProtoEntityLayout.AttrRawDataOff, 8));
        if (!ProtoEntityLayout.IsPlausiblePtr(bsPtr))
            return (attrId, Array.Empty<byte>());

        var bsBlob = pm.ReadBytes(bsPtr, ProtoEntityLayout.ByteStrLengthOff + 4);
        if (bsBlob is null || bsBlob.Length < ProtoEntityLayout.ByteStrLengthOff + 4)
            return (attrId, Array.Empty<byte>());

        ulong arrPtr = BinaryPrimitives.ReadUInt64LittleEndian(
            bsBlob.AsSpan(ProtoEntityLayout.ByteStrObjOff, 8));
        int start = BinaryPrimitives.ReadInt32LittleEndian(
            bsBlob.AsSpan(ProtoEntityLayout.ByteStrStartOff, 4));
        int length = BinaryPrimitives.ReadInt32LittleEndian(
            bsBlob.AsSpan(ProtoEntityLayout.ByteStrLengthOff, 4));
        if (length < 0 || length > ProtoEntityLayout.MaxRawAttrLen
            || !ProtoEntityLayout.IsPlausiblePtr(arrPtr))
            return (attrId, Array.Empty<byte>());

        ulong dataAddr = arrPtr + (ulong)ProtoEntityLayout.ArrayDataOff + (ulong)Math.Max(0, start);
        var raw = pm.ReadBytes(dataAddr, length);
        return (attrId, raw ?? Array.Empty<byte>());
    }

    /// <summary>
    /// Read a full Entity at <paramref name="entityAddr"/>. Returns null when
    /// the header itself can't be read; otherwise returns a snapshot. With
    /// <paramref name="decodeAttrs"/>=false only header fields are populated.
    /// <paramref name="attrFilter"/>, when non-null, restricts varint decode
    /// to those attribute ids (raw_attrs still records every id).
    /// </summary>
    public static EntitySnapshot? ReadEntityFull(
        IMemorySource pm,
        ulong entityAddr,
        bool decodeAttrs = true,
        IReadOnlySet<int>? attrFilter = null)
    {
        var body = pm.ReadBytes(entityAddr, 0x40);
        if (body is null || body.Length < ProtoEntityLayout.EntityAttrsOff + 8)
            return null;

        long uuid = BinaryPrimitives.ReadInt64LittleEndian(
            body.AsSpan(ProtoEntityLayout.EntityUuidOff, 8));
        int entType = BinaryPrimitives.ReadInt32LittleEndian(
            body.AsSpan(ProtoEntityLayout.EntityEntTypeOff, 4));
        ulong attrsColl = BinaryPrimitives.ReadUInt64LittleEndian(
            body.AsSpan(ProtoEntityLayout.EntityAttrsOff, 8));

        var attrs = new Dictionary<int, int>();
        var rawAttrs = new List<(ulong, int)>();

        if (decodeAttrs && attrsColl != 0 && ProtoEntityLayout.IsPlausiblePtr(attrsColl))
        {
            var attrPtrs = ReadRepeatedFieldAttrs(pm, attrsColl);
            foreach (var attrPtr in attrPtrs)
            {
                var pair = ReadAttrIdAndRawBytes(pm, attrPtr);
                if (pair is null) continue;
                var (attrId, raw) = pair.Value;
                rawAttrs.Add((attrPtr, attrId));
                if (attrFilter is not null && !attrFilter.Contains(attrId)) continue;
                if (raw.Length == 0) continue;
                attrs[attrId] = DecodeVarintI32(raw);
            }
        }

        return new EntitySnapshot(entityAddr, uuid, entType, attrsColl, attrs, rawAttrs);
    }
}
